assets/app/scripts/services/data.js

```diff
@@ -559,7 +559,31 @@ angular.module('openshiftConsole')
     }
   };
 
+  function itemKindForList(data) {
+    var pos, kind = data.kind;
+    if (kind) {
+      pos = kind.lastIndexOf("List");
+      if (pos === kind.length - 4)
+        return kind.substring(0, pos);
+    }
+    return null;
+  }
+
   DataService.prototype._listOpComplete = function(type, context, data) {
+
+    /*
+     * Here we normalize all items to have a kind property.
```

Comment:
Kubernetes returns data inconsistently on resources. Sometimes ```.kind``` is set, and sometimes not. This depends on the mechanism used to retrieve the item (```GET``` a list vs. ```GET``` a resource vs. ```WATCH```). Here we normalize them before anything else in the app gets their hands on them.

Without this fix I have to recreate all the objects before passing them to the topology. (ie: make a shallow copy of the resource, and set a kind on it). Anything in the app that will deal with resources generically will need to know their kind. This data should be carried on the resource itself.

I think the correct fix is to cope with Kubernetes inconsistencies as close to the source as possible.

But like I said there is a work around if you insist, (shallow copy all resources before passing them to the topology view ... and in the future anything else that deals with resources generically). The work around is inefficient and hacky, but I can do it. Please confirm.